File: project_euler/python/problem4.py
"""
Largest palindrome product
Problem 4

A palindromic number reads the same both ways. 
The largest palindrome made from the product 
of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
---------------------------------------------------------------------

STATUS: COMPLETE
ANSWER: 906609
RUNTIME: 1.0s
COMPLEXITY: O(n * len(max_number))
OPERATION COUNT: 4*(10**6)
"""

def is_palindrome(num): # O(len(num)), C(5*len(num))
    return str(num) == str(num)[::-1]

# SOLUTION
three_dig_nums = [n for n in range(100, 1000)] # 900 numbers
pals = []
for index1 in range(len(three_dig_nums)): # O()
    for index2 in range(index1, len(three_dig_nums)):
        num1 = three_dig_nums[index1]
        num2  = three_dig_nums[index2]
        if is_palindrome(num1 * num2):
            pals.append(num1 * num2)
print(max(pals))

File: project_euler/java/problem4.java
/*
Largest palindrome product
Problem 4

A palindromic number reads the 
same both ways. The largest palindrome 
made from the product of two 2-digit 
numbers is 9009 = 91 × 99.

Find the largest palindrome made from 
the product of two 3-digit numbers.

SOLUTION: 906609
*/

public class problem4 {

	public static boolean is_palindrome(int num) {
		// returns true if num is a palindrome
		// returns false otherwise
		String s_num = "" + num; // convert num to a string
		int k = s_num.length(); 
		int i = 0;
		while (i <= k-i) {
			if (s_num.charAt(i) != s_num.charAt(k-1-i)) return false;
			i++;
		}
		return true;
	}

	public static int max_palindrome_product(int digs) {
		// returns the largest palindrome which is
		// the product of two number with digs digits
		int max_pal = 0;
		for (int num1 = (int) Math.pow(10, digs-1); num1 < (int) Math.pow(10, digs); num1++) {
			for (int num2 = num1; num2 < (int) Math.pow(10, digs); num2++) {
				if (is_palindrome(num1 * num2) && num1 * num2 > max_pal) {
					max_pal = num1 * num2;
				}
			}
		}
		return max_pal;
	}

	public static void main(String[] args) {
		// TESTS 1
		// System.out.println(""+123+" palindromicity "+is_palindrome(123));
		// System.out.println(""+1231+" palindromicity "+is_palindrome(1231));
		// System.out.println(""+12321+" palindromicity "+is_palindrome(12321));
		// System.out.println(""+12312+" palindromicity "+is_palindrome(12312));
		// System.out.println(""+12399+" palindromicity "+is_palindrome(12399));
		// System.out.println(""+32123+" palindromicity "+is_palindrome(32123));
		// System.out.println(""+123+" palindromicity "+is_palindrome(123));
		// System.out.println(""+0+" palindromicity "+is_palindrome(0))
		System.out.println(max_palindrome_product(1));
		System.out.println(max_palindrome_product(2));
		System.out.println(max_palindrome_product(3));
	}
}